File: user-frontend/src/types/order.ts
import { Product } from "./product";

export type OrderStatus = 'pending' | 'processing' | 'shipped' | 'delivered' | 'cancelled';

export interface OrderItem {
    id: number;
    product_id: number;
    product_name: string;
    quantity: number;
    price: number;
    product: Product;
}

export interface Order {
    id: number;
    user_id: number;
    user_email: string;
    user_name?: string; 
    shipping_full_name: string;
    status: OrderStatus;
    total_price: number;
    items: OrderItem[];
    shipping_address: string;
    tracking_number?: string;
    created_at: string;
}

export interface OrderStatusData {
  order_id: number;
  status: OrderStatus;
  tracking_number?: string;
}

File: backend/app/db/models/order.py
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, Float
from sqlalchemy.orm import relationship
from datetime import datetime
from app.db.base import Base

class Order(Base):
    __tablename__ = "orders"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"))
    status = Column(String, default="pending")
    created_at = Column(DateTime, default=datetime.utcnow)

    shipping_full_name = Column(String, nullable=False)
    shipping_address = Column(String, nullable=False)
    shipping_city = Column(String, nullable=False)
    shipping_postal_code = Column(String, nullable=False)
    shipping_phone = Column(String, nullable=False)

    tracking_number = Column(String, nullable=True)

    user = relationship("User", back_populates="orders")
    items = relationship("OrderItem", back_populates="order")
    
    @property
    def total_price(self) -> float:
        return sum(item.price * item.quantity for item in self.items)


class OrderItem(Base):
    __tablename__ = "order_items"

    id = Column(Integer, primary_key=True, index=True)
    order_id = Column(Integer, ForeignKey("orders.id"))
    product_id = Column(Integer, ForeignKey("products.id"))
    quantity = Column(Integer, nullable=False)
    price = Column(Float, nullable=False)  # price per unit at order time

    order = relationship("Order", back_populates="items")
    product = relationship("Product")